### PR TITLE
Fix #11540: Enable asan on Brave

### PR DESCRIPTION
**Pull request by jumde**
Resolves https://github.com/brave/brave-browser/issues/11540

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [x] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

1. Build with `--is_asan` flag
2. Run:

```
npm run test -- brave_unit_tests  --filter=ToolsSanityTest.DISABLED_AddressSanitizerLocalOOBCrashTest --run_disabled_tests
```

and verify you see an asan crash.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: build/commands/lib/config.js [hunk 1]
```diff
@@ -183,7 +186,7 @@ Config.prototype.buildArgs = function () {
     enable_widevine: true,
     target_cpu: this.targetArch,
     is_official_build: this.isOfficialBuild(),
-    is_debug: this.isDebug(),
+    is_debug: this.isDebug() && !this.buildAsan(),
```
Comment (bridiver) [inline build/commands/lib/config.js:189]: I don't think this is necessary, this is enforced by gn

File: build/commands/lib/config.js [hunk 2]
```diff
@@ -168,6 +168,9 @@ Config.prototype.buildArgs = function () {
   const chrome_version_parts = this.chromeVersion.split('.')
 
   let args = {
+    is_asan: this.buildAsan(),
```
Comment (bridiver) [inline build/commands/lib/config.js:171]:
```
if (this.isAsan) {
   args.is_asan = ...
```

File: build/commands/lib/config.js [hunk 3]
```diff
@@ -470,6 +480,11 @@ Config.prototype.update = function (options) {
     this.targetOS = options.target_os
   }
 
+
+  if (options.asan) {
```
Comment (bridiver) [inline build/commands/lib/config.js:484]: please keep the option names consistent with chromium `asan` -> `is_asan`

**Review by iefremov — APPROVED** [review]

**Review by bsclifton — APPROVED** [review]
LGTM! 😄

File: build/commands/scripts/commands.js [hunk 4]
```diff
@@ -191,6 +192,7 @@ program
   .option('--test_launcher_jobs <test_launcher_jobs>', 'Number of jobs to launch', parseInt, '4')
   .option('--target_os <target_os>', 'target OS')
   .option('--target_arch <target_arch>', 'target architecture')
+  .option('--run_disabled', 'run disabled tests')
```
Comment (bridiver) [inline build/commands/scripts/commands.js:195]: `run_disabled_tests` for consistency

**Review by bridiver — APPROVED** [review]

File: components/brave_ads/test/BUILD.gn [hunk 5]
```diff
@@ -67,6 +66,17 @@ source_set("brave_ads_unit_tests") {
       "//brave/vendor/bat-native-ads/src/bat/ads/internal/url_util_unittest.cc",
     ]
 
+    if (!is_asan) {
+      # Move to the sources outside the !is_asan check when
+      # https://github.com/brave/brave-browser/issues/11891
```
Comment (tmancey) [inline components/brave_ads/test/BUILD.gn:71]: Fixed

File: components/brave_ads/test/BUILD.gn [hunk 6]
```diff
@@ -67,6 +66,17 @@ source_set("brave_ads_unit_tests") {
...
+      # https://github.com/brave/brave-browser/issues/11892
```
Comment (tmancey) [inline components/brave_ads/test/BUILD.gn:72]: Resolved as part of 11891

**Review by tmancey — CHANGES_REQUESTED** [review]
Fixed issues related to ads, can you please rebase once merged so we can remove the checks? Thanks

**Review by tmancey — APPROVED** [review]
LGTM++

**Review by NejcZdovc — APPROVED** [review]

**Pull request merged**